### DISCUSSION TITLE
Using make to interact with containers

### DISCUSSION THREAD

**Discussion by Guikingone** [dagger/dagger#2786] [General]
Hi 👋🏻 

First, a firm handshake to the team behind Dagger, you're doing a great job easing our CI's 😅 

This discussion is focused on how Dagger works and how it may be adapted to specific needs, I've been through the documentation and the API (both `dagger.io` and `universe.dagger.io`) and I didn't found anything about using Dagger as an orchestrator on top of tools like `Makefile` / `Taskfile`, did I miss something? 

Here's the use case I'm facing: 

- The project use Docker / Docker-Compose but has specific configuration regarding public URL's / domains thanks to `Caddy`.
- `Cypress` is used to launch / write E2E tests while querying a public URL published by `Caddy` (something like `foo.localhost`).
- Each commands / utilities are launched via `Makefile` and dedicated recipes.

So, the fast solution could be to use `Dagger` to use the default images / containers and orchestrate them using the plan, thing is, some commands / services are orchestrated via the `Makefile` and I prefer to keep using things like `make up` to launch the containers rather than relying on `Dagger` (as I can control which services are launched, etc) but I didn't find anything about using commands.

I've been through the steps like `Bash` and so on but it seems that every time, it requires to launch a dedicated container 🙁

So, the question is: Could `Dagger` launch "os commands / tools" like `Makefile` or should I "rewrite" steps via the `Dagger` API?

Thanks for the help and have a great day 🙂

**Comment by TomChv** [comment]
Hello 👋 

Thanks for your message!

>  Could Dagger launch "os commands / tools" like Makefile or should I "rewrite" steps via the Dagger API?

Every Dagger's instruction are done in containers so you'll need to use [`client`](https://docs.dagger.io/1203/client) to interact with your host's filesystem.
But yes, you can create actions as you would do in a Makefile.

For an example, so our [own Dagger's CI](https://github.com/dagger/dagger/blob/main/ci.cue), you'll see that we defined a set of action, for instance : 
```
 build  
 test   
 lint   
 gitpod 
```

Same as a Makefile, you can execute `build` action with `dagger do build`. An action can be almost everything, you just need to write it with Dagger API or wrap Makefile in Dagger action. It depends of you.
